en: add doc for multi-cluster prometheus monitor (#1775)
pingcap/docs-tidb-operator: en/deploy-tidb-monitor-across-multiple-kubernetes.md

---
title: Monitor a TiDB Cluster across Multiple Kubernetes Clusters

> summary:

Author: ```suggestion
summary: Learn how to monitor a TiDB cluster across multiple Kubernetes clusters and aggregate the monitoring data into the common Prometheus multi-cluster monitoring systems.
```

        # "cluster" label has been used by the TiDB metrics already.
        # For more information, please refer to the issue
        # https://github.com/pingcap/tidb-operator/issues/4219.
        k8s_clsuter: ${kubernetes_cluster_name}
        # add other meta labels here
        #region: us-east-1
      initializer:
        baseImage: pingcap/tidb-monitor-initializer
        version: v5.4.0
      persistent: true
      storage: 100Gi
      storageClassName: ${storageclass_name}
      prometheus:
        baseImage: prom/prometheus
        logLevel: info
        remoteWrite:
        - url: ${remote_write_url}
        retentionTime: 2h
        version: v2.27.1
      reloader:
        baseImage: pingcap/tidb-monitor-reloader
        version: v1.0.1
      imagePullPolicy: IfNotPresent
    EOF
    ```

## Pull data from Prometheus

The pull method pulls monitoring data from Prometheus instances in different Kubernetes clusters and aggregates the data into a global view. This section describes how to pull data [using Thanos Query](#using-thanos-query) and [using Prometheus Federation](#using-prometheus-federation).

<SimpleTab>
<div label="Thanos Query">

### Using Thanos Query

> In the example of this section, one Thanos Sidecar is deployed for each Prometheus (`TidbMonitor`) component. The Thanos query component is used for aggregation queries. If you do not need long-term storage, you can skip the deployment of some components such as thanos-store and S3.

Reviewer: ```suggestion
In the example of this section, one Thanos Sidecar is deployed for each Prometheus (`TidbMonitor`) component. The thanos-query component is used for aggregation queries. If you do not need long-term storage, you can skip the deployment of some components such as thanos-store and S3.
```

        baseImage: pingcap/tidb-monitor-initializer
        version: v5.4.0
      persistent: true
      storage: 20Gi
      storageClassName: ${storageclass_name}
      prometheus:
        baseImage: prom/prometheus
        logLevel: info
        version: v2.27.1
      reloader:
        baseImage: pingcap/tidb-monitor-reloader
        version: v1.0.1
      thanos:
        baseImage: quay.io/thanos/thanos
        version: v0.22.0
        #enable config below if long-term storage is needed.
        #objectStorageConfig:
        #  key: objectstorage.yaml
        #  name: thanos-objectstorage
      imagePullPolicy: IfNotPresent
    EOF
    ```

3. Configure Thanos Query Stores:

    You can specify the store nodes by the static service discovery method. In the Thanos Query command line's starting parameters, add `--store=${cluster_name}-prometheus.${cluster_namespace}.svc.${cluster_domain}:10901` to specify the store node. Replace the variable with the actual value.

    If you use other service discovery methods, refer to [thanos-service-discovery](https://thanos.io/tip/thanos/service-discovery.md).

</div>
<div label="Federation">

### Using Prometheus Federation

> In the example of this section, the Federation Prometheus server is used as an entry point for unified storage and query. This method is recommended for small data volumes.

Reviewer 2: ```suggestion
In the example of this section, the Federation Prometheus server is used as an entry point for unified storage and query. This method is only considered for small data volumes.
```
we actually do not recommend using this.

#### Configure Federation Prometheus

For details on Federation, refer to [Federation](https://prometheus.io/docs/prometheus/latest/federation/#hierarchical-federation). After the deployment of Prometheus, you need to modify the Prometheus configuration, and add the host information of the Prometheus (`TidbMonitor`) to be aggregated.

```yaml
scrape_configs:
  - job_name: 'federate'
    scrape_interval: 15s

    honor_labels: true
    metrics_path: '/federate'

    params:
      'match[]':
        - '{__name__=~".+"}'

    static_configs:
      - targets:
        - 'source-prometheus-1:9090'
        - 'source-prometheus-2:9090'
        - 'source-prometheus-3:9090'
```

</div>
</SimpleTab>

## Visualize monitoring data using Grafana

After collecting data using Prometheus, you can visualize multi-cluster monitoring data using Grafana.

1. Obtain the Grafana Dashboards related to TiDB components by running the following command:

    {{< copyable "shell-regular" >}}

    ```shell
    # set tidb version here
    version=nightly
    docker run --rm -i -v ${PWD}/dashboards:/dashboards/ pingcap/tidb-monitor-initializer:${version} && \
    cd dashboards
    ```

    > **Note:**
    >

>     > In the command above, `${version}` is the version of the Initializer image, which should be consistent with the TiDB version. However, currently only the `nightly` version of the Initializer image supports **multiple Kubernetes clusters** monitoring, so this example uses the `nightly` image.

Reviewer 2: since TiDB 6.0 has been released, maybe change the example to v6.0.0 here?

Author: Sure. It's your call.

Author: Updated in [5b7f056](https://github.com/pingcap/docs-tidb-operator/pull/1775/commits/5b7f0568c0c5ff373ae35cb15fd08b77843db841)

    ```shell
    # set tidb version here
    version=v6.1.0
    docker run --rm -i -v ${PWD}/dashboards:/dashboards/ pingcap/tidb-monitor-initializer:${version} && \
    cd dashboards
    ```

    > **Note:**
    >

>     > In the command above, `${version}` is the version of the Initializer image, which should be consistent with the TiDB version. Currently only `v6.0.0` and later versions of the Initializer image support **multiple Kubernetes clusters** monitoring.

Reviewer 3: ```suggestion
    > In the command above, `${version}` is the version of the Initializer image, which should be consistent with the TiDB version. Currently, only `v6.0.0` and later versions of the Initializer image support **multiple Kubernetes clusters** monitoring.
```
